File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    genre_album = {}
    genre_song = {}
    
    for i in range(len(genres)):
        k = genres[i]
        if k in genre_album:
            genre_album[k] += plays[i]
            genre_song[k].append((i, plays[i]))
        else:
            genre_album[k] = plays[i]
            genre_song[k] = [(i, plays[i])]
    
    genre_album = dict(sorted(genre_album.items(), key=lambda x: x[1], reverse=True))
    
    answer = []
    for g in genre_album:
        genre_songs = sorted(genre_song[g], key=lambda x: (-x[1], x[0]))
        if len(genre_songs) == 1:
            answer.append(genre_songs[0][0])
        else:
            for i in range(2):
                answer.append(genre_songs[i][0])
        
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;
import java.util.stream.Collectors;

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        HashMap<String, Integer> genres_play = new HashMap<>();
        HashMap<String, List<List<Integer>>> genres_songs = new HashMap<>();
        
        for (int i = 0; i < genres.length; i++) {
            String k = genres[i];
            if (genres_play.containsKey(k)) {
                genres_play.put(k, genres_play.get(k) + plays[i]);
                genres_songs.get(k).add(Arrays.asList(i, plays[i]));
            } else {
                genres_play.put(k, plays[i]);
                genres_songs.put(k, new ArrayList<>());
                genres_songs.get(k).add(Arrays.asList(i, plays[i]));
            }
        }
        
        List<String> sorted_genre = genres_play.entrySet()
            .stream()
            .sorted((e1, e2) -> e2.getValue().compareTo(e1.getValue())) 
            .map(Map.Entry::getKey)
            .collect(Collectors.toList());
        
        List<Integer> answer = new ArrayList<>();
        for(String g : sorted_genre) {
            List<List<Integer>> songs = genres_songs.get(g);
            songs.sort(
                Comparator.comparing((List<Integer> l) -> l.get(1)).reversed()  
                          .thenComparing(l -> l.get(0))                       
            );

            
            if (songs.size() == 1) {
                answer.add(songs.get(0).get(0));
                continue;
            } 
            
            for (int i = 0; i < 2; i++) {
                answer.add(songs.get(i).get(0));
            }
        }
        return answer.stream().mapToInt(Integer::intValue).toArray();
    }
}